Amplify/Core/Support/AmplifyInProcessReportingOperation.swift

```diff
@@ -0,0 +1,96 @@
+//
+// Copyright 2018-2020 Amazon.com,
+// Inc. or its affiliates. All Rights Reserved.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+import Foundation
+
+/// An AmplifyOperation that emits InProcess values intermittently during the operation.
+///
+/// Unlike a regular `AmplifyOperation`, which emits a single Result at the completion of the operation's work, an
+/// `AmplifyInProcessReportingOperation` may emit intermediate values while its work is ongoing. These values could be
+/// incidental to the operation (such as a `Storage.downloadFile` operation reporting Progress values periodically as
+/// the download proceeds), or they could be the primary delivery mechanism for an operation (such as a
+/// `GraphQLSubscriptionOperation`'s emitting new subscription values).
+open class AmplifyInProcessReportingOperation<
+    Request: AmplifyOperationRequest,
+    InProcess,
+    Success,
+    Failure: AmplifyError
+>: AmplifyOperation<Request, Success, Failure> {
+    var inProcessListenerUnsubscribeToken: UnsubscribeToken?
+    var secondaryResultListenerToken: UnsubscribeToken?
+
+    public init(categoryType: CategoryType,
+                eventName: HubPayloadEventName,
+                request: Request,
+                inProcessListener: InProcessListener? = nil,
+                resultListener: ResultListener? = nil) {
+
+        super.init(categoryType: categoryType, eventName: eventName, request: request, resultListener: resultListener)
+
+        // If the inProcessListener is present, we need to register a hub event listener for it, and ensure we
+        // automatically unsubscribe when we receive a completion event for the operation
+        if let inProcessListener = inProcessListener {
+            self.inProcessListenerUnsubscribeToken = subscribe(inProcessListener: inProcessListener)
+        }
+    }
+
+    /// Registers an in-process listener for this operation. If the operation completes, this listener will
+    /// automatically be removed.
+    ///
+    /// - Parameter inProcessListener: The listener for in-process events
+    /// - Returns: an UnsubscribeToken that can be used to remove the listener from Hub
+    func subscribe(inProcessListener: @escaping InProcessListener) -> UnsubscribeToken {
+        let channel = HubChannel(from: categoryType)
+        let filterById = HubFilters.forOperation(self)
+
+        var progressListenerToken: UnsubscribeToken!
+        let progressHubListener: HubListener = { payload in
```

Comment:
`inProcess` vs `progress`, I was thinking whether this should called `AmplifyProgressOperation`, i don't have a strong opinion on this, just that I noticed here we define a local variable `progressListenerToken` and not something like `inProcessListenerToken`. At the same time, InProcess makes more sense for Subscriptions values. but end of the day all of it is all typealias'd

Reply:
I considered "progress" as a generic term, but it's somewhat overloaded because of the Foundation type, and the fact that we explicitly use Progress as an "in-process" value. That made me settle on the slightly more awkward but more generically applicable "in-process". I can be argued out of "in-process" to something else, but I feel like "progress" or variations thereof are going to cause more confusion than they'll solve.